src/Compilers/Core/Portable/Compilation/CompilationOptions.cs

```diff
@@ -198,6 +198,11 @@ public abstract class CompilationOptions
         /// </summary>
         public ImmutableDictionary<string, ReportDiagnostic> SpecificDiagnosticOptions { get; protected set; }
 
+        /// <summary>
+        /// Warning report option for each warning.
+        /// </summary>
+        public ImmutableDictionary<SyntaxTree, ImmutableDictionary<string, ReportDiagnostic>> PerTreeDiagnosticOptions { get; protected set; }
```

Comment:
>public ImmutableDictionary<SyntaxTree, ImmutableDictionary<string, ReportDiagnostic>> PerTreeDiagnosticOptions { get; protected set; } [](start = 8, length = 134)

I don't think this information belongs in compilation options. Syntax trees are not part of options they can be added and removed from the compilation at any time and now that would have to be coordinated with value of this property. I think it is better to attach this information to the SyntaxTree directly, perhaps through ParseOptions. There would be no need to build dictionaries, all relevant pieces are together, added and removed together.